static/scripts/assignment_table.js

```diff
@@ -0,0 +1,482 @@
+                row.loadQ2 = 0;
+
+                //This line is only used to store course assignments, so the other values are empty.
+                const emptyKeys = ['org', 'promoter', 'totalLoad', 'loadQ1', 'loadQ2'];
+                emptyKeys.forEach(key => {
+                    emptyRow[key] = "";
+                });
+
+                const userPrefs = Object.values(preferences).filter(pref => pref.researcher_id === researcher.id);
+
+                let pos = 1;
+                courses.forEach(course => {
+                    const isPref = userPrefs.find(pref => pref.course_id === course.id);
+                    const code = course.code;
+                    row[code] = isPref ? pos++ : "";
+                    emptyRow[code] = "";
+                });
+
+                rows.push(row);
+                rows.push(emptyRow);
+            }
+            return rows;
+        }
+
+        const fixedRows = fixedRowsData();
+        const userRows = userRowsData();
+
+
+        function getCourseColumns() {
+            const fixedColumns = [
+                {data: 'researchers.id'},
+                {data: 'researchers.name'},
+                {data: 'org'},
+                {data: 'promoter'},
+                {data: 'totalLoad'},
+                {data: 'loadQ1'},
+                {data: 'loadQ2'}
+            ];
+
+            courses.forEach(course => {
+                const code = course.code;
+                let col = {data: code};
+                fixedColumns.push(col)
+            });
+            return fixedColumns;
+        }
+
+        const columns = getCourseColumns();
+
+        let data = fixedRows.concat(userRows);
+        const nbrLines = data.length - 1;
+        const nbrCols = columns.length - 1;
+
+        const mergeCellsSettings = [];
+        for (let row = lenFixedRowsText; row < nbrLines; row += 2) {
+            for (let col = 1; col <= lenFixedHeaders - 1; col++) {
+                mergeCellsSettings.push({
+                    row: row,
+                    col: col,
+                    rowspan: 2,
+                    colspan: 1
+                });
+            }
+        }
+
+        function storeDataLocally(data) {
+            localStorage.setItem('table', JSON.stringify(data));
+        }
+
+        function storeCellsMeta(cellsMeta) {
+            localStorage.setItem('cellsMeta', JSON.stringify(cellsMeta));
+        }
+
+        function retrieveDataLocally() {
+            const storedData = localStorage.getItem('table');
+            if (storedData) {
+                data = JSON.parse(storedData);
+            }
+        }
+
+        function retrieveCellsMetaLocally() {
+            const cellsMeta = localStorage.getItem('cellsMeta');
+            return cellsMeta ? JSON.parse(cellsMeta) : null;
+        }
+
+        function resetDataLocally() {
+            localStorage.clear();
+        }
+
+        // Verify if the table is modified locally
+        retrieveDataLocally();
```

Comment:
Please either:

- remove the function definition as it is used once and alter the value of a variable out of its scope (`data`)
- make the function return the loaded value.